### DISCUSSION TITLE
Alternative to click types

### DISCUSSION THREAD

**Discussion by Mitmischer** [janluke/cloup#73] [Q&A] [answered]
In click, I could specify the type of an option, like this: [1](https://click.palletsprojects.com/en/8.0.x/arguments/?highlight=click%20path#file-arguments) with the types specified here: [2](https://click.palletsprojects.com/en/8.0.x/api/#types) [3](https://click.palletsprojects.com/en/8.0.x/parameters/?highlight=click%20path#parameter-types) .

This is expecially useful for files with the `click.File` class.
`cloup.File` does not exist; Also it looks to me as if the whole typing functionality from click was not adopted. Are there alternatives to achieve similar functionality in cloup?

**Comment by janluke** [comment, marked as answer]
**UPDATE(v0.10.0):** Click types are now re-exported by Cloup, so you don't need to import Click.

------------

Cloup is meant to be used in tandem with Click, it's not a replacement for Click. You can use Click types as you do with Click:

```python
import click
import cloup

@cloup.command()
@cloup.option('-o', '--out', type=click.File(...))
def f(out):
    ...
```

More in general, what you can do with Click, you can do with Cloup too, because Cloup commands *are* Click commands (subclasses) and Cloup parameters *are* Click parameters.

Note that, even though Cloup has Click as a dependency, you should explicitly add Click to your requirements too because you are gonna use it directly in your code (usually only for types). I'll add this to the installation instructions.

**Comment by janluke** [comment]
Update: in v0.10.0, Click types are now re-exported by Cloup, so you don't need to import Click.

**Comment by Mitmischer** [comment]
Sounds good, thanks for the quick response!
Yeah, it was a bit unclear to me whether cloup wanted to be a drop-in-replacement for click or not. With the types re-exported, cloup looks like a proper drop-in-replacement (is this correct?)
Out of curiosity, have you considered merging cloup back into click?

**Reply by janluke** [reply]
There's still stuff from click that cloup doesn't export. See the "installation" section on the docs.

Cloup adds features that click maintainers don't want to add, except better type hints.

**Reply by Mitmischer** [reply]
Hmm, quite sad that click maintainers don't want to see this added.
